Prevent empty statements at the end of --eval or --loop code

WebDrake opened the pull request:
The `--eval` and `--loop` flags allow the user to specify small snippets of code that are placed inside the body of an auto-generated `main` that is written to a file and evaluated.

The existing implementation uses two separate string concatenations for the two different cases.  Besides the code duplication, this means that the resulting code generation cannot be tested.

A further more subtle issue arises from the existing implementation's insistence on adding a closing `;` to the generated code.  While this takes care of the careless user who forgets to write one themselves, it means that if the user does close their statements correctly, we end up with an empty statement on the end of the code to be evaluated.  While some D compilers don't care about this, some object: https://github.com/dlang/tools/pull/297#issuecomment-362447334

... meaning that `rdmd` will fail to successfully evaluate the code.

These patches rework the implementation around a new `makeEvalCode` function taking two parameters: the `string[]` of code snippets to be evaluated, and a boolean flag indicating whether or not the code should be treated as the body of a loop (in other words, whether the code snippets derived from an `--eval` or a `--loop` flag).

Generation of the innermost code (the common part for both `--eval` and `--loop` options) is implemented in a second new function, `innerEvalCode`, which takes care of joining the different entries in the `string[]`, appending a terminating `;` only if one is not already present.  In consequence, the generated code should never end in a blank statement (at least, not unless it was explicitly put there by the user).

Unittests have been provided with both the new functions to ensure that they produce the expected outcomes.  The first patch provides the functions themselves; the second uses them to replace the old implementation.

WebDrake commented:
For reference: the new function should produce identical results to the old ad-hoc implementations, apart from the trailing `;`.

WebDrake commented on rdmd.d:
```diff
@@ -839,6 +835,103 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
+    string code = eval.join("\n");
+    if (stripRight(eval[$ - 1])[$ - 1] != ';')
+        code ~= ';';
+    yap("[inner-code]\n", code);
```
Oh, whoops.  This shouldn't be there, it was a debugging check.  I'll remove.

WebDrake commented:
Removed the unintended debug statement:

```diff
diff --git a/rdmd.d b/rdmd.d
index d87ac76..118471c 100755
--- a/rdmd.d
+++ b/rdmd.d
@@ -855,7 +855,6 @@ string innerEvalCode(string[] eval)
     string code = eval.join("\n");
     if (stripRight(eval[$ - 1])[$ - 1] != ';')
         code ~= ';';
-    yap("[inner-code]\n", code);
     return code;
 }
```

marler8997 commented on rdmd.d:
```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
+        to the inner code of either the program or
+        the loop
+*/
+string makeEvalCode(string[] eval, bool loop = false)
```
`Flag!"loop"` instead?

marler8997 commented on rdmd.d:
```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
+{
+    import std.string: join, stripRight;
+    string code = eval.join("\n");
+    if (stripRight(eval[$ - 1])[$ - 1] != ';')
```
Should you handle the case where `eval.length` == 0, or the case where `eval[$-1].length == 0` rather than causing an overflow?

marler8997 commented on rdmd.d:
```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
 ";
 
+/**
+    Joins together the code provided via an `--eval`
```
I believe the agreed upon convention is to have the ddoc content aligned with the comment, which in this case means no leading whitespace.

wilzbach replied:
Yes, please follow the [DStyle](https://dlang.org/dstyle.html#phobos_documentation) for new code.

marler8997 commented on rdmd.d:
```diff
@@ -180,18 +180,14 @@ int main(string[] args)
                 "program file ('" ~ args[programPos] ~ "').");
-        root = makeEvalFile(importWorld ~ "void main(char[][] args) {\n"
-                ~ std.string.join(eval, "\n") ~ ";\n}");
+        root = makeEvalFile(makeEvalCode(eval));
```
`makeEvalCode(eval, No.loop)` seems better

marler8997 commented on rdmd.d:
```diff
@@ -180,18 +180,14 @@ int main(string[] args)
-                ~ "foreach (line; std.stdio.stdin.byLine()) {\n"
-                ~ std.string.join(loop, "\n")
-                ~ ";\n} }");
+        root = makeEvalFile(makeEvalCode(loop, true));
```
`makeEvalCode(eval, Yes.loop)` seems better

wilzbach commented on rdmd.d:
```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
+*/
+string innerEvalCode(string[] eval)
+{
+    import std.string: join, stripRight;
```
Nit: space before and after `:` for selective imports (DStyle for official code).

wilzbach commented on rdmd.d:
```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
+        ~ "void main(char[][] args) {%s%s\n%s}";
+
+    immutable innerCodeOpening =
+        loop ? " foreach (line; std.stdio.stdin.byLine()) {\n"
```
Idiomatic D would just use `.byLine` - you could use this opportunity to change this bit.

wilzbach commented on rdmd.d:
```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
+{
+    import std.format : format;
+    immutable codeFormat = importWorld
+        ~ "void main(char[][] args) {%s%s\n%s}";
```
Let's use this opportunity and correct the mutability here - it should have been `void main(string args)...`

wilzbach commented on rdmd.d:
```diff
@@ -839,6 +835,102 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
+                  innerCodeClosing);
+}
+
+unittest
```
Could be `@safe`

WebDrake replied:
`@safe` won't go with the use of `assumeSafeAppend` inside `innerEvalCode`.  That might be possible to drop, but OTOH, I'm not sure I care to overly restrict how anyone edits or revises these functions in future.  It's not like `rdmd.d` makes common use of these kinds of constraints.

WebDrake commented:
@wilzbach I don't think it makes sense to bundle changes to how the code is generated, with changes to _what_ code is generated (beyond the base goal of fixing the trailing-empty-statement issue).

I don't want to assume that there were not deliberate motivations behind the choices made here, and I don't want to risk breaking anyone's use-case by changing things that do not need to be changed.

WebDrake commented:
I've updated the ddoc, replaced the `bool` with a `Flag!"loop"` (good call @marler8997!), and reworked the internals of `innerEvalCode` to avoid the risk of overflow.  A small by-product is that it now right-strips trailing whitespace from the generated code (meaning unittests have been updated slightly), but this makes for no practical difference in outcome.

Overall diff:

```diff
diff --git a/rdmd.d b/rdmd.d
index 118471c..250f3f2 100755
--- a/rdmd.d
+++ b/rdmd.d
@@ -17,7 +17,7 @@
 import std.algorithm, std.array, core.stdc.stdlib, std.datetime,
     std.digest.md, std.exception, std.file, std.getopt,
     std.parallelism, std.path, std.process, std.range, std.regex,
-    std.stdio, std.string, std.typetuple;
+    std.stdio, std.string, std.typecons, std.typetuple;
 
 version (Posix)
 {
@@ -180,14 +180,14 @@ int main(string[] args)
     {
         enforce(programPos == args.length, "Cannot have both --loop and a " ~
                 "program file ('" ~ args[programPos] ~ "').");
-        root = makeEvalFile(makeEvalCode(loop, true));
+        root = makeEvalFile(makeEvalCode(loop, Yes.loop));
         argsBeforeProgram ~= "-d";
     }
     else if (eval.ptr)
     {
         enforce(programPos == args.length, "Cannot have both --eval and a " ~
                 "program file ('" ~ args[programPos] ~ "').");
-        root = makeEvalFile(makeEvalCode(eval));
+        root = makeEvalFile(makeEvalCode(eval, No.loop));
         argsBeforeProgram ~= "-d";
     }
     else if (programPos < args.length)
@@ -836,24 +836,24 @@ import std.stdio, std.algorithm, std.array, std.ascii, std.base64,
 ";
 
 /**
-    Joins together the code provided via an `--eval`
-    or `--loop` flag, ensuring a trailing `;` is
-    added if not already provided by the user
-
-    Params:
-        eval = array of strings generated by the
-               `--eval` or `--loop` rdmd flags
-
-    Returns:
-        string of code to be evaluated, corresponding
-        to the inner code of either the program or
-        the loop
+Joins together the code provided via an `--eval` or `--loop`
+flag, ensuring a trailing `;` is added if not already provided
+by the user
+
+Params:
+    eval = array of strings generated by the `--eval`
+           or `--loop` rdmd flags
+
+Returns:
+    string of code to be evaluated, corresponding to the
+    inner code of either the program or the loop
 */
 string innerEvalCode(string[] eval)
 {
-    import std.string: join, stripRight;
-    string code = eval.join("\n");
-    if (stripRight(eval[$ - 1])[$ - 1] != ';')
+    import std.string : join, stripRight;
+    // assumeSafeAppend just to avoid unnecessary reallocation
+    string code = eval.join("\n").stripRight.assumeSafeAppend;
+    if (code.length > 0 && code[$ - 1] != ';')
         code ~= ';';
     return code;
 }
@@ -864,34 +864,33 @@ unittest
     assert(innerEvalCode([`writeln("Hello!");`]) == `writeln("Hello!");`);
 
     // test with trailing whitespace
-    assert(innerEvalCode([`writeln("Hello!")  `]) == `writeln("Hello!")  ;`);
-    assert(innerEvalCode([`writeln("Hello!");  `]) == `writeln("Hello!");  `);
+    assert(innerEvalCode([`writeln("Hello!")  `]) == `writeln("Hello!");`);
+    assert(innerEvalCode([`writeln("Hello!");  `]) == `writeln("Hello!");`);
 
     // test with multiple entries
     assert(innerEvalCode([`writeln("Hello!");  `, `writeln("You!")  `])
-           == "writeln(\"Hello!\");  \nwriteln(\"You!\")  ;");
+           == "writeln(\"Hello!\");  \nwriteln(\"You!\");");
     assert(innerEvalCode([`writeln("Hello!");  `, `writeln("You!"); `])
-           == "writeln(\"Hello!\");  \nwriteln(\"You!\"); ");
+           == "writeln(\"Hello!\");  \nwriteln(\"You!\");");
 }
 
 /**
-    Formats the code provided via `--eval` or `--loop`
-    flags into a string of complete program code that
-    can be written to a file and then compiled
-
-    Params:
-        eval = array of strings generated by the
-               `--eval` or `--loop` rdmd flags
-        loop = set to `true` if this code comes
-               from a `--loop` flag, `false` (the
-               default) otherwise
-
-    Returns:
-        string of code to be evaluated, corresponding
-        to the inner code of either the program or
-        the loop
+Formats the code provided via `--eval` or `--loop` flags into a
+string of complete program code that can be written to a file
+and then compiled
+
+Params:
+    eval = array of strings generated by the `--eval` or
+           `--loop` rdmd flags
+    loop = set to `Yes.loop` if this code comes from a
+           `--loop` flag, `No.loop` if it comes from an
+           `--eval` flag
+
+Returns:
+    string of code to be evaluated, corresponding to the
+    inner code of either the program or the loop
 */
-string makeEvalCode(string[] eval, bool loop = false)
+string makeEvalCode(string[] eval, Flag!"loop" loop)
 {
     import std.format : format;
     immutable codeFormat = importWorld
@@ -914,21 +913,21 @@ unittest
     // innerEvalCode already tests the cases for different
     // contents in `eval` array, so let's focus on testing
     // the difference based on the `loop` flag
-    assert(makeEvalCode([`writeln("Hello!") `]) ==
+    assert(makeEvalCode([`writeln("Hello!") `], No.loop) ==
            importWorld
            ~ "void main(char[][] args) {\n"
-           ~ "writeln(\"Hello!\") ;\n}");
+           ~ "writeln(\"Hello!\");\n}");
 
-    assert(makeEvalCode([`writeln("What!"); `], false) ==
+    assert(makeEvalCode([`writeln("What!"); `], No.loop) ==
            importWorld
            ~ "void main(char[][] args) {\n"
-           ~ "writeln(\"What!\"); \n}");
+           ~ "writeln(\"What!\");\n}");
 
-    assert(makeEvalCode([`writeln("Loop!") ; `], true) ==
+    assert(makeEvalCode([`writeln("Loop!") ; `], Yes.loop) ==
            importWorld
            ~ "void main(char[][] args) { "
            ~ "foreach (line; std.stdio.stdin.byLine()) {\n"
-           ~ "writeln(\"Loop!\") ; \n} }");
+           ~ "writeln(\"Loop!\") ;\n} }");
 }
 
 string makeEvalFile(string todo)
```

wilzbach commented:
> I don't want to assume that there were not deliberate motivations behind the choices made here, and I don't want to risk breaking anyone's use-case by changing things that do not need to be changed.

Fair enough, 

> Overall diff:

Are you aware of these two handy features from GH?

https://patch-diff.githubusercontent.com/raw/dlang/tools/pull/303.diff
https://github.com/dlang/tools/pull/303?w=1

WebDrake commented:
> Are you aware of these two handy features from GH?

Nice, but if I understand correctly, that's the overall diff of the PR.  I posted the diff with between the original and versions of the PR. ;-)

WebDrake commented:
Ping on this? Is there anything that remains to do (e.g. a changelog entry or Bugzilla issue)?
